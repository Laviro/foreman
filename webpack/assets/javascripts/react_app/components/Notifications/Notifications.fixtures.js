/* eslint-disable camelcase */
/* eslint-disable camelcase */
import Immutable from 'seamless-immutable';

export const initialState = Immutable({
  expandedGroup: null,
  isDrawerOpen: null,
  hasUnreadMessages: false,
});

export const panelRequest = {
  group: 'Hosts',
};
export const NotificationRequest = {
  id: 51435,
};

export const notifications = [
  {
    id: 52435,
    seen: false,
    level: 'info',
    text: 'some.example.com has been deleted successfully',
    created_at: '2017-04-17T17:29:12.664Z',
    group: 'Hosts',
    actions: {},
  },
  {
    id: 51435,
    seen: false,
    level: 'info',
    text: 'notified successfully',
    created_at: '2017-04-17T17:29:12.664Z',
    group: 'Hosts',
    actions: {},
  },
];

export const stateWithNotifications = Immutable({
  isDrawerOpen: true,
  expandedGroup: 'Hosts',
  hasUnreadMessages: true,
  notifications,
});

export const componentMountData = { url: '/notification_recipients' };

export const serverResponse = `{"data": { "notifications":[
  {"id":1,"seen":true,"level":"info","text":"notification1","created_at":"2017-02-23T05:00:28.715Z",
  "group":"React devs","actions":{}},
  {"id":2,"seen":false,"level":"info","text":"notification2","created_at":"2017-02-23T05:00:28.715Z",
  "group":"React devs","actions":{}}]}}`;
