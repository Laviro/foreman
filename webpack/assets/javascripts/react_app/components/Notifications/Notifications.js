import onClickOutside from 'react-onclickoutside';
import React from 'react';
import PropTypes from 'prop-types';
import {
  NotificationDrawerWrapper,
  NotificationDrawerPanelWrapper,
} from 'patternfly-react';
import { noop, translateObject } from '../../common/helpers';
import ToggleIcon from './components/ToggleIcon';
import { NOTIFICATIONS } from '../../redux/consts';
import './notifications.scss';

class Notifications extends React.Component {
  componentDidMount() {
    const {
      startNotificationsPolling,
      data: { url },
    } = this.props;

    startNotificationsPolling(url);
  }

  handleClickOutside() {
    const { isDrawerOpen, isReady, toggleDrawer } = this.props;

    if (isReady && isDrawerOpen) {
      toggleDrawer();
    }
  }

  componentWillUnmount() {
    const { stopPolling } = this.props;
    stopPolling(NOTIFICATIONS);
  }

  render() {
    const {
      notifications,
      isDrawerOpen,
      toggleDrawer,
      expandGroup,
      expandedGroup,
      markAsRead,
      markGroupAsRead,
      clearNotification,
      clearGroup,
      hasUnreadMessages,
      isReady,
      clickedLink,
      translations,
    } = this.props;

    const notificationGroups = Object.entries(notifications).map(
      ([key, group]) => ({
        panelkey: key,
        panelName: key,
        notifications: group,
      })
    );

    return (
      <div>
        <ToggleIcon
          hasUnreadMessages={hasUnreadMessages}
          onClick={toggleDrawer}
        />
        {isReady && isDrawerOpen && (
          <NotificationDrawerWrapper
            panels={notificationGroups}
            expandedPanel={expandedGroup}
            togglePanel={expandGroup}
            onNotificationAsRead={markAsRead}
            onNotificationHide={clearNotification}
            onMarkPanelAsRead={markGroupAsRead}
            onMarkPanelAsClear={clearGroup}
            onClickedLink={clickedLink}
            toggleDrawerHide={toggleDrawer}
            isExpandable={false}
            translations={translateObject(translations)}
          />
        )}
      </div>
    );
  }
}

Notifications.propTypes = {
  data: PropTypes.shape({
    url: PropTypes.string.isRequired,
  }).isRequired,
  isDrawerOpen: PropTypes.bool,
  isReady: PropTypes.bool,
  notifications: PropTypes.object,
  expandedGroup: PropTypes.string,
  hasUnreadMessages: PropTypes.bool,
  clickedLink: PropTypes.func,
  startNotificationsPolling: PropTypes.func,
  toggleDrawer: PropTypes.func,
  expandGroup: PropTypes.func,
  markAsRead: PropTypes.func,
  markGroupAsRead: PropTypes.func,
  clearNotification: PropTypes.func,
  clearGroup: PropTypes.func,
  stopPolling: PropTypes.func,
  translations: PropTypes.shape({
    title: PropTypes.string,
    unreadEvent: PropTypes.string,
    unreadEvents: PropTypes.string,
    emptyState: PropTypes.string,
    readAll: PropTypes.string,
    clearAll: PropTypes.string,
    deleteNotification: PropTypes.string,
  }),
};

Notifications.defaultProps = {
  isDrawerOpen: false,
  isReady: false,
  notifications: {},
  expandedGroup: null,
  hasUnreadMessages: false,
  clickedLink: noop,
  startNotificationsPolling: noop,
  toggleDrawer: noop,
  expandGroup: noop,
  markAsRead: noop,
  markGroupAsRead: noop,
  clearNotification: noop,
  clearGroup: noop,
  stopPolling: noop,
  translations: NotificationDrawerPanelWrapper.defaultProps.translations,
};

export default onClickOutside(Notifications);
