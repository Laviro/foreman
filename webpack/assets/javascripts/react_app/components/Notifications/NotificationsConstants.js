export const NOTIFICATIONS = 'NOTIFICATIONS';
export const NOTIFICATIONS_TOGGLE_DRAWER = 'NOTIFICATIONS_TOGGLE_DRAWER';
export const NOTIFICATIONS_SET_EXPANDED_GROUP =
  'NOTIFICATIONS_SET_EXPANDED_GROUP';
export const NOTIFICATIONS_MARK_AS_READ = 'NOTIFICATIONS_MARK_AS_READ';
export const NOTIFICATIONS_MARK_GROUP_AS_READ =
  'NOTIFICATIONS_MARK_GROUP_AS_READ';
export const NOTIFICATIONS_MARK_AS_CLEAR = 'NOTIFICATIONS_MARK_AS_CLEAR';
export const NOTIFICATIONS_MARK_GROUP_AS_CLEARED =
  'NOTIFICATIONS_MARK_GROUP_AS_CLEARED';
export const NOTIFICATIONS_POLLING_STARTED = 'NOTIFICATIONS_POLLING_STARTED';
export const NOTIFICATIONS_LINK_CLICKED = 'NOTIFICATIONS_LINK_CLICKED';
