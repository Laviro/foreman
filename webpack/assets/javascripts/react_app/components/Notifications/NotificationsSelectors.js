import { groupBy } from 'lodash';

export const selectNotificationsNamespace = state => state.notifications;
export const selectIsDrawerOpen = state =>
  selectNotificationsNamespace(state).isDrawerOpen;
export const selectExpandedGroup = state =>
  selectNotificationsNamespace(state).expandedGroup;
export const selectHasUnreadMessages = state =>
  selectNotificationsNamespace(state).hasUnreadMessages;
export const selectNotifications = state => {
  const { notifications } = selectNotificationsNamespace(state);
  return groupBy(notifications, n => n.group);
};
export const selectIsReady = state => !!selectNotifications(state);
