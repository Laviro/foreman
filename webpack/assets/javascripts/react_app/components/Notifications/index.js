import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import * as NotificationActions from './NotificationsActions';
import reducer from './NotificationsReducer';
import { APIActions } from '../../redux/API';
import {
  selectIsDrawerOpen,
  selectNotifications,
  selectExpandedGroup,
  selectIsReady,
  selectHasUnreadMessages,
} from './NotificationsSelectors';
import Notifications from './Notifications';

// map state to props
const mapStateToProps = state => ({
  isDrawerOpen: selectIsDrawerOpen(state),
  notifications: selectNotifications(state),
  expandedGroup: selectExpandedGroup(state),
  isReady: selectIsReady(state),
  hasUnreadMessages: selectHasUnreadMessages(state),
});

// map action dispatchers to props
const mapDispatchToProps = dispatch =>
  bindActionCreators({ ...NotificationActions, ...APIActions }, dispatch);

// export reducers
export const reducers = { notifications: reducer };

// export connected component
export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Notifications);
