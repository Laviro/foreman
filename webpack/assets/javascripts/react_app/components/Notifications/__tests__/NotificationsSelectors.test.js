import { testSelectorsSnapshotWithFixtures } from 'react-redux-test-utils';
import {
  selectNotifications,
  selectIsDrawerOpen,
  selectIsReady,
  selectHasUnreadMessages,
  selectExpandedGroup,
} from '../NotificationsSelectors';
import { stateWithNotifications } from '../Notifications.fixtures';

const state = stateWithNotifications;

const fixtures = {
  'should return Notifications': () => selectNotifications(state),
  'should return Notifications isDrawerOpen': () => selectIsDrawerOpen(state),
  'should return Notifications expandedGroup': () => selectExpandedGroup(state),
  'should return Notifications hasUnreadMessages': () =>
    selectHasUnreadMessages(state),
  'should return Notifications isReady': () => selectIsReady(state),
};

describe('Notifications selectors', () =>
  testSelectorsSnapshotWithFixtures(fixtures));
