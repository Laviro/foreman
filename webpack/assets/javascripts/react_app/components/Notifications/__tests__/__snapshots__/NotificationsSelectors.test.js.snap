// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notifications selectors should return Notifications 1`] = `Object {}`;

exports[`Notifications selectors should return Notifications expandedGroup 1`] = `undefined`;

exports[`Notifications selectors should return Notifications hasUnreadMessages 1`] = `undefined`;

exports[`Notifications selectors should return Notifications isDrawerOpen 1`] = `undefined`;

exports[`Notifications selectors should return Notifications isReady 1`] = `true`;
